fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/Scopes.java

```diff
@@ -79,33 +91,177 @@ public QuantifiedValue resolveIdentifier(@Nonnull final String identifier) {
-            this.parentScope = parentScope;
-            this.boundQuantifiers = ImmutableMap.copyOf(boundQuantifiers);
-            this.graphExpansionBuilder = graphExpansionBuilder;
-        }
+        private final ImmutableList<Column<? extends Value>> projectionList;
 
         @Nullable
-        public Scope getParentScope() {
-            return parentScope;
+        private final QueryPredicate predicate;
+
+        @Nonnull
+        private final ImmutableList<QuantifiedValue> quantifiedValues;
+
+        Scope(@Nonnull final ImmutableMap<CorrelationIdentifier, Quantifier> quantifiers, @Nonnull final ImmutableList<Column<? extends Value>> projectionList, @Nonnull ImmutableList<QuantifiedValue> quantifiedValues, @Nullable final QueryPredicate predicate) {
+            this.quantifiers = quantifiers;
+            this.projectionList = projectionList;
+            this.predicate = predicate;
+            this.quantifiedValues = quantifiedValues;
         }
 
-        public Map<CorrelationIdentifier, QuantifiedValue> getBoundQuantifiers() {
-            return boundQuantifiers;
+        @Nonnull
+        public SelectExpression convertToSelectExpression() {
+            GraphExpansion.Builder builder = GraphExpansion.builder();
+            builder.addAllQuantifiers(quantifiers.values().asList())
+                    .addAllResultColumns(projectionList)
+                    .addAllResultValues(quantifiedValues);
+            if (predicate != null) {
+                builder.addPredicate(predicate);
+            }
+            return builder.build().seal().buildSelect();
         }
 
         @Nonnull
-        public GraphExpansion.Builder getGraphExpansionBuilder() {
-            return graphExpansionBuilder;
+        public Optional<Quantifier> getQuantifier(@Nonnull final String alias) {
+            return getQuantifier(CorrelationIdentifier.of(alias));
+        }
+
+        @Nonnull
+        public Optional<Quantifier> getQuantifier(@Nonnull final CorrelationIdentifier alias) {
+            return Optional.ofNullable(this.quantifiers.get(alias));
+        }
+
+        public boolean hasQuantifier(@Nonnull final String alias) {
+            return hasQuantifier(CorrelationIdentifier.of(alias));
+        }
+
+        public boolean hasQuantifier(@Nonnull final CorrelationIdentifier alias) {
+            return quantifiers.containsKey(alias);
+        }
+
+        /**
+         * a builder of a {@link Scope}.
+         */
+        public static class Builder {
+
+            @Nullable
+            private Builder parent;
+
+            @Nonnull
+            private final ImmutableMap.Builder<CorrelationIdentifier, Quantifier> quantifiers;
+
+            @Nullable
+            private QueryPredicate predicate;
+
+            @Nonnull
+            private ImmutableList.Builder<QuantifiedValue> quantifiedValues;
+
+            @Nonnull
+            private ImmutableList.Builder<Column<? extends Value>> projectionList;
```

Comment:
So for relational stuff this is fine since you will always return a record that is fed list of stuff. But this doesn't allow you to put the direct `QuantifiedObjectValue` value into the select expression you'll eventually create. The same problem existed in `GraphExpansion`. I solved it in the join enumeration PR.